### PR TITLE
Android updates

### LINKED ISSUE #5: Set up a Continuous Integration server
Key functionality
- Verify that all PRs build successfully
- Report the output APK size
- Ideally, report app installation time on a Moto G4 Play

I believe reporting that information at the pull request might require a bot. For an initial version, it's OK if we have to go look at the CI output.

### PR DESCRIPTION
**Pull request by magneticflux-**
Closes #5. Travis CI still needs to be enabled on the OpenFTC Organization and on this particular repository.

---
A lot of work needs to be done for #15.
#4 should be only a few lines now that the build has been updated to Android Gradle plugin 3.0.0.

Request changes if there are any issues!

### CODE REVIEW

File: Blocks/src/main/java/com/google/blocks/AbstractProgrammingModeActivity.java [hunk 1]
```diff
@@ -33,11 +33,10 @@
  * @author lizlooney@google.com (Liz Looney)
  */
 public abstract class AbstractProgrammingModeActivity extends ThemedActivity {
+    private final Timer timer = new Timer();
```
Comment (qwertychouskie) [inline Blocks/src/main/java/com/google/blocks/AbstractProgrammingModeActivity.java]: Spacing issue

Comment (qwertychouskie) [reply Blocks/src/main/java/com/google/blocks/AbstractProgrammingModeActivity.java]: Also why does this need to be moved?

File: Blocks/src/main/java/com/google/blocks/AbstractProgrammingModeActivity.java [hunk 2]
```diff
@@ -179,7 +178,7 @@ public void run() {
    * typeface.
    */
   private void setTypeface() {
-    Typeface typeface = Typeface.create(Typeface.SERIF, 0);
+      Typeface typeface = Typeface.create(Typeface.SERIF, Typeface.NORMAL);
```
Comment (qwertychouskie) [inline Blocks/src/main/java/com/google/blocks/AbstractProgrammingModeActivity.java]: Spacing again

File: Blocks/src/main/java/com/google/blocks/ftcrobotcontroller/runtime/BlocksOpMode.java [hunk 3]
```diff
@@ -44,12 +44,10 @@
   private static final String LOG_PREFIX = "BlocksOpMode - ";
 
   private static final AtomicReference<String> fatalErrorMessageHolder = new AtomicReference<String>();
-
-  private static Activity activity;
-  private static WebView webView;
   private static final AtomicReference<String> nameOfOpModeLoadedIntoWebView = new AtomicReference<String>();
   private static final Map<String, Access> javascriptInterfaces = new ConcurrentHashMap<String, Access>();
-
+    private static Activity activity;
```
Comment (qwertychouskie) [inline Blocks/src/main/java/com/google/blocks/ftcrobotcontroller/runtime/BlocksOpMode.java]: Here too

File: Blocks/src/main/java/com/google/blocks/ftcrobotcontroller/runtime/BlocksOpMode.java [hunk 4]
```diff
@@ -67,6 +65,61 @@
     logPrefix = LOG_PREFIX + "\"" + project + "\" - ";
   }
 
+    /**
```
Comment (qwertychouskie) [inline Blocks/src/main/java/com/google/blocks/ftcrobotcontroller/runtime/BlocksOpMode.java]: And this whole section

File: Blocks/src/main/java/com/google/blocks/ftcrobotcontroller/runtime/BlocksOpMode.java [hunk 5]
```diff
@@ -282,16 +335,17 @@ public void run() {
     }
   }
 
-  private void addJavascriptInterfaces(HardwareItemMap hardwareItemMap) {
-    addJavascriptInterfacesForIdentifiers();
-    addJavascriptInterfacesForHardware(hardwareItemMap);
+    @SuppressLint("JavascriptInterface")
```
Comment (qwertychouskie) [inline Blocks/src/main/java/com/google/blocks/ftcrobotcontroller/runtime/BlocksOpMode.java]: And here

File: Blocks/src/main/java/com/google/blocks/ftcrobotcontroller/runtime/BlocksOpMode.java [hunk 6]
```diff
@@ -282,16 +335,17 @@ public void run() {
...
 
-    for (Map.Entry<String, Access> entry : javascriptInterfaces.entrySet()) {
-      String identifier = entry.getKey();
-      Access access = entry.getValue();
-      webView.addJavascriptInterface(access, identifier);
+        for (Map.Entry<String, Access> entry : javascriptInterfaces.entrySet()) {
+            String identifier = entry.getKey();
+            Access access = entry.getValue();
+            webView.addJavascriptInterface(access, identifier);
+        }
```
Comment (qwertychouskie) [inline Blocks/src/main/java/com/google/blocks/ftcrobotcontroller/runtime/BlocksOpMode.java]: And here

File: Blocks/src/main/java/com/google/blocks/ftcrobotcontroller/runtime/BlocksOpMode.java [hunk 7]
```diff
@@ -494,57 +521,30 @@ private void clearScript() {
...
 
-  /**
-   * Sets the {@link WebView} so that all BlocksOpModes can access it.
-   */
-  public static void setActivityAndWebView(Activity a, WebView wv) {
-    if (activity == null && webView == null) {
-      addOpModeRegistrar();
-    }
+    private class BlocksOpModeAccess extends Access {
+        private final Object scriptFinishedLock;
```
Comment (qwertychouskie) [inline Blocks/src/main/java/com/google/blocks/ftcrobotcontroller/runtime/BlocksOpMode.java]: And here

File: Blocks/src/main/java/com/google/blocks/ftcrobotcontroller/runtime/BlocksOpMode.java [hunk 8]
```diff
@@ -494,57 +521,30 @@ private void clearScript() {
...
-        // configuration, there will be a console message like this:
-        // "ReferenceError: left_drive is not defined".
-        String message = consoleMessage.message();
-        if (message.startsWith(REFERENCE_ERROR_PREFIX)) {
-          RobotLog.e(LOG_PREFIX + "fatalErrorMessage: " + message);
-          fatalErrorMessageHolder.compareAndSet(null, message);
+        private BlocksOpModeAccess(String identifier, Object scriptFinishedLock, AtomicBoolean scriptFinished) {
+            super(BlocksOpMode.this, identifier);
+            this.scriptFinishedLock = scriptFinishedLock;
+            this.scriptFinished = scriptFinished;
```
Comment (qwertychouskie) [inline Blocks/src/main/java/com/google/blocks/ftcrobotcontroller/runtime/BlocksOpMode.java]: And here

File: Blocks/src/main/java/com/google/blocks/ftcrobotcontroller/runtime/BlocksOpMode.java [hunk 9]
```diff
@@ -494,57 +521,30 @@ private void clearScript() {
...
-          List<OpModeMeta> projects = ProjectsUtil.fetchEnabledProjectsWithJavaScript();
-          for (OpModeMeta opModeMeta : projects) {
-            manager.register(opModeMeta, new BlocksOpMode(opModeMeta.name));
-          }
-        } catch (Exception e) {
-          RobotLog.logStackTrace(e);
+        @SuppressWarnings("unused")
+        @JavascriptInterface
+        public void scriptStarting() {
+            RobotLog.i(getLogPrefix() + "scriptStarting");
```
Comment (qwertychouskie) [inline Blocks/src/main/java/com/google/blocks/ftcrobotcontroller/runtime/BlocksOpMode.java]: And here

File: Blocks/src/main/java/com/google/blocks/ftcrobotcontroller/runtime/BlocksOpMode.java [hunk 10]
```diff
@@ -494,57 +521,30 @@ private void clearScript() {
...
-    RobotLog.w(BlocksOpMode.class.getSimpleName(), "registerAll(OpModeManager) is deprecated and will be removed soon, as calling it is unnecessary in this and future API version");
+        @SuppressWarnings("unused")
+        @JavascriptInterface
+        public void scriptFinished() {
+            RobotLog.i(getLogPrefix() + "scriptFinished");
+            synchronized (scriptFinishedLock) {
+                scriptFinished.set(true);
+                scriptFinishedLock.notifyAll();
+            }
+    }
```
Comment (qwertychouskie) [inline Blocks/src/main/java/com/google/blocks/ftcrobotcontroller/runtime/BlocksOpMode.java]: Here too

File: FtcCommon/src/main/res/layout/lynx_module_configure_address_spin_item.xml [hunk 11]
```diff
@@ -1,11 +1,11 @@
 <?xml version="1.0" encoding="utf-8"?>
-<CheckedTextView android:id="@android:id/text1"
+<CheckedTextView xmlns:android="http://schemas.android.com/apk/res/android"
+    android:id="@android:id/text1"
```
Comment (qwertychouskie) [inline FtcCommon/src/main/res/layout/lynx_module_configure_address_spin_item.xml]: Also wrong spacing

File: RobotCore/build.gradle [hunk 12]
```diff
@@ -18,9 +16,8 @@ android {
             proguardFiles getDefaultProguardFile('proguard-android.txt'), 'proguard-rules.pro'
         }
     }
-
 }
 
 dependencies {
-    compile fileTree(include: ['*.jar'], dir: 'libs') //mark as api once we update the Android plugin to 3.0+
+    api fileTree(include: ['*.jar'], dir: 'libs') //mark as api once we update the Android plugin to 3.0+
```
Comment (qwertychouskie) [inline RobotCore/build.gradle]: You can probably remove the comment here now

File: build.gradle [hunk 13]
```diff
@@ -5,10 +5,11 @@
  */
 buildscript {
     repositories {
+        google()
```
Comment (qwertychouskie) [inline build.gradle]: Is this extra repo really needed?

**Review by qwertychouskie — CHANGES_REQUESTED** [review]
A few nitpicks as follows:

**Comment by Windwoes** [comment]
Changing the targetSdk version to 26 will cause loads of crashes because >=23 will enable runtime permissions

File: Blocks/src/main/java/com/google/blocks/AbstractProgrammingModeActivity.java (hunk 1 above)
Comment (magneticflux-) [reply Blocks/src/main/java/com/google/blocks/AbstractProgrammingModeActivity.java]: I accidentally ran my formatter on changed files, I'll undo the spacing and order changes.

**Comment by magneticflux-** [comment]
@FROGbots-4634 Per the [documentation](https://developer.android.com/training/permissions/requesting.html):
>If the device is running Android 5.1 or lower, or your app's target SDK is 22 or lower: If you list a dangerous permission in your manifest, the user has to grant the permission when they install the app; if they do not grant the permission, the system does not install the app at all.

>If the device is running Android 6.0 or higher, and your app's target SDK is 23 or higher: The app has to list the permissions in the manifest, and it must request each dangerous permission it needs while the app is running. The user can grant or deny each permission, and the app can continue to run with limited capabilities even if the user denies a permission request.

Are there teams that are running on a phone with API >= 23? The phones my team had were all API 19 or 21. I believe the official FTC documentation says that phones above Android Lollipop (API 21) aren't supported.

File: build.gradle (hunk 13 above)
Comment (magneticflux-) [reply build.gradle]: It's required for Android Gradle plugin 3.0.0

**Review by qwertychouskie — APPROVED** [review]
Looking good!

**Comment by Windwoes** [comment]
@magneticflux- 
1. Yes, there are teams running >= API 23. In fact, the phone officially recommend by FIRST for this year is running Marshmallow.

2. Android Lollipop is not supported due to a USB OTG bug that only allows it to communicate with one downstream USB device at a time. This was fixed in Marshmallow.

3. Regarding >= 21 not being supported, that's quite incorrect. In fact, per the game manual for this year, the ZTE Speed is the only one of the allowed list of phones that is not *required* to be on Marshmallow or higher.

**Comment by magneticflux-** [comment]
@FROGbots-4634 What about requesting Camera permissions on app launch? That would prevent any Vuforia errors from being denied the camera. Are there any other permission errors?

**Comment by Windwoes** [comment]
@magneticflux- You'd also need to request external storage permission.

Requesting permissions at all would add a lot of complexity to MainActivity, something we do not want to do. The goal of OpenFTC is to add features as unobtrusively as possible, modifying as little of FIRST's code as possible. Any significant changes to MainActivity will make any future merges from FIRST's upstream very complicated and just a pain in general.

**Comment by magneticflux-** [comment]
@FROGbots-4634 I added a permission request for Camera and Read/Write External Storage in about 5 lines of code. It doesn't give any user-friendly info, but it does allow permissions.

Additionally, the app was compiled with API 23 before I updated it to API 26. That means that the permission changes were in place before I increased the API further.

**Comment by Windwoes** [comment]
@magneticflux- incorrect, compileSdk is different from targetSdk.

You can have a compileSdk of 26 and still target 19.

Before, it was compiled with 23 but targeted 19

**Comment by magneticflux-** [comment]
@FROGbots-4634 Thanks, I missed that detail!

The next issue to be tackled is the `.jar` dependencies of `RobotCore`, namely the `javac.jar` dependency. @NoahAndrews What was your reasoning behind including the entire Java compiler?

**Comment by Windwoes** [comment]
@magneticflux- He didn't include it - FIRST did. It's needed to compile user code written with OnBotJava

**Review by ItsHarper — CHANGES_REQUESTED** [review]
Sorry this took me so long to get to. 

I'll merge this, but I'd like you to add documentation to the readme that this will require Android Studio 3.0 and updated SDK components. I'd really like to see screenshots included of what the errors will look like if you try to build this in Android Studio 2.3.x or with an Android SDK that doesn't have platform 26 installed.

**Comment by ItsHarper** [comment]
Thank you for submitting this, but I'm going to close it now. We ended up moving to gradle plugin 3 in a different commit, and the Travis stuff can wait for now.

**Pull request closed without merging**